File: packages/gatekeeper/src/search-index.ts
type JSONObject = Record<string, unknown>;

export default class SearchIndex {
    private docs = new Map<string, JSONObject>();
    private static readonly ARRAY_WILDCARD_END = /\[\*]$/;
    private static readonly ARRAY_WILDCARD_MID = /\[\*]\./;

    store(did: string, doc: object): void {
        this.docs.set(did, JSON.parse(JSON.stringify(doc)) as JSONObject);
    }

    get(did: string): object | null {
        const v = this.docs.get(did);
        return v ? JSON.parse(JSON.stringify(v)) : null;
    }

    delete(did: string): void {
        this.docs.delete(did);
    }

    clear(): void {
        this.docs.clear();
    }

    get size(): number {
        return this.docs.size;
    }

    searchDocs(q: string): string[] {
        const out: string[] = [];
        for (const [did, doc] of this.docs.entries()) {
            if (JSON.stringify(doc).includes(q)) out.push(did);
        }
        return out;
    }

    queryDocs(where: Record<string, unknown>): string[] {
        const entry = Object.entries(where)[0] as [string, unknown] | undefined;
        if (!entry) {
            return [];
        }
        const [rawPath, cond] = entry;
        if (typeof cond !== 'object' || cond === null || !Array.isArray((cond as { $in?: unknown }).$in)) {
            throw new Error('Only {$in:[...]} supported');
        }
        const list = (cond as { $in: unknown[] }).$in;

        const isKeyWildcard = rawPath.endsWith('.*');
        const isValueWildcard = rawPath.includes('.*.');
        const isArrayTail = SearchIndex.ARRAY_WILDCARD_END.test(rawPath);
        const isArrayMid = SearchIndex.ARRAY_WILDCARD_MID.test(rawPath);

        const result: string[] = [];

        for (const [did, doc] of this.docs.entries()) {
            let match = false;

            if (isArrayTail) {
                const basePath = rawPath.replace(SearchIndex.ARRAY_WILDCARD_END, '');
                const arr = this.getPath(doc, basePath);
                if (Array.isArray(arr)) {
                    match = arr.some(v => list.includes(v));
                }
            } else if (isArrayMid) {
                const [prefix, suffix] = rawPath.split('[*].');
                const arr = this.getPath(doc, prefix);
                if (Array.isArray(arr)) {
                    match = arr.some(el => list.includes(this.getPath(el, suffix)));
                }
            } else if (isKeyWildcard) {
                const basePath = rawPath.slice(0, -2);
                const obj = this.getPath(doc, basePath);
                if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
                    const keys = Object.keys(obj as Record<string, unknown>);
                    match = keys.some(k => list.includes(k));
                }
            } else if (isValueWildcard) {
                const [prefix, suffix] = rawPath.split('.*.');
                const obj = this.getPath(doc, prefix);
                if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
                    const values = Object.values(obj as Record<string, unknown>);
                    match = values.some(v => list.includes(this.getPath(v, suffix)));
                }
            } else {
                const val = this.getPath(doc, rawPath);
                match = list.includes(val);
            }

            if (match) {
                result.push(did);
            }
        }

        return result;
    }

    private getPath(root: unknown, path: string): unknown {
        if (!path || root == null) {
            return undefined;
        }

        const clean = path.startsWith('$.') ? path.slice(2) : path.startsWith('$') ? path.slice(1) : path;
        if (!clean) {
            return root;
        }

        const parts = clean.split('.');

        let cur: unknown = root;
        for (const rawPart of parts) {
            if (cur == null) {
                return undefined;
            }

            const idx = Number.isInteger(+rawPart) ? +rawPart : null;

            if (idx !== null && Array.isArray(cur)) {
                cur = cur[idx];
                continue;
            }

            if (typeof cur === 'object') {
                cur = (cur as Record<string, unknown>)[rawPart];
            } else {
                return undefined;
            }
        }
        return cur;
    }
}

File: rust/services/gatekeeper/src/search_index.rs
use std::collections::HashMap;

use serde_json::Value;

#[derive(Default)]
pub(crate) struct SearchIndex {
    docs: HashMap<String, serde_json::Map<String, Value>>,
    order: Vec<String>,
}

impl SearchIndex {
    pub(crate) fn store(&mut self, did: &str, doc: &Value) {
        let data = doc
            .get("didDocumentData")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_default();
        if !self.docs.contains_key(did) {
            self.order.push(did.to_string());
        }
        self.docs.insert(did.to_string(), data);
    }

    pub(crate) fn delete(&mut self, did: &str) {
        if self.docs.remove(did).is_some() {
            self.order.retain(|entry| entry != did);
        }
    }

    pub(crate) fn clear(&mut self) {
        self.docs.clear();
        self.order.clear();
    }

    pub(crate) fn size(&self) -> usize {
        self.docs.len()
    }

    pub(crate) fn search_docs(&self, q: &str) -> Vec<String> {
        let mut out = Vec::new();
        for did in &self.order {
            let Some(doc) = self.docs.get(did) else {
                continue;
            };
            if serde_json::to_string(doc)
                .map(|value| value.contains(q))
                .unwrap_or(false)
            {
                out.push(did.clone());
            }
        }
        out
    }

    pub(crate) fn query_docs(&self, where_clause: &Value) -> anyhow::Result<Vec<String>> {
        let Some((raw_path, cond)) = where_clause.as_object().and_then(|map| map.iter().next()) else {
            return Ok(Vec::new());
        };
        let list = cond
            .get("$in")
            .and_then(Value::as_array)
            .ok_or_else(|| anyhow::anyhow!("Only {{$in:[...]}} supported"))?;

        let mut result = Vec::new();
        for did in &self.order {
            let Some(doc) = self.docs.get(did) else {
                continue;
            };
            if query_match(&Value::Object(doc.clone()), raw_path, list) {
                result.push(did.clone());
            }
        }
        Ok(result)
    }
}

fn query_match(root: &Value, raw_path: &str, list: &[Value]) -> bool {
    if let Some(base_path) = raw_path.strip_suffix("[*]") {
        return json_path_get(root, base_path)
            .and_then(Value::as_array)
            .map(|arr| arr.iter().any(|value| list.contains(value)))
            .unwrap_or(false);
    }

    if let Some((prefix, suffix)) = raw_path.split_once("[*].") {
        return json_path_get(root, prefix)
            .and_then(Value::as_array)
            .map(|arr| {
                arr.iter()
                    .filter_map(|item| json_path_get(item, suffix))
                    .any(|value| list.contains(value))
            })
            .unwrap_or(false);
    }

    if let Some(base_path) = raw_path.strip_suffix(".*") {
        return json_path_get(root, base_path)
            .and_then(Value::as_object)
            .map(|obj| obj.keys().any(|key| list.contains(&Value::String(key.clone()))))
            .unwrap_or(false);
    }

    if let Some((prefix, suffix)) = raw_path.split_once(".*.") {
        return json_path_get(root, prefix)
            .and_then(Value::as_object)
            .map(|obj| {
                obj.values()
                    .filter_map(|item| json_path_get(item, suffix))
                    .any(|value| list.contains(value))
            })
            .unwrap_or(false);
    }

    json_path_get(root, raw_path)
        .map(|value| list.contains(value))
        .unwrap_or(false)
}

fn json_path_get<'a>(root: &'a Value, path: &str) -> Option<&'a Value> {
    if path.is_empty() {
        return Some(root);
    }

    let clean = path
        .strip_prefix("$.")
        .or_else(|| path.strip_prefix('$'))
        .unwrap_or(path);
    if clean.is_empty() {
        return Some(root);
    }

    let mut current = root;
    for raw_part in clean.split('.') {
        if let Ok(index) = raw_part.parse::<usize>() {
            current = current.as_array()?.get(index)?;
        } else {
            current = current.get(raw_part)?;
        }
    }
    Some(current)
}
